python/paddle/fluid/incubate/fleet/base/fleet_base.py

```diff
@@ -0,0 +1,308 @@
+#   Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import abc
+import sys
+
+from enum import Enum
+
+from paddle.fluid.optimizer import Optimizer
+
+from role_maker import RoleMakerBase, Role
+from role_maker import MPISymetricRoleMaker
+from role_maker import UserDefinedRoleMaker
+
+
+class Mode(Enum):
+    TRANSPILER = 1,
+    PSLIB = 2,
+    COLLECTIVE = 3
+
+
+class Fleet(object):
+    """
+    Fleet is the base class, transpiler and pslib are implementation of Fleet.
+    """
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, mode):
+        """
+        Init Fleet with mode, mode must be an instance of Mode.
+
+        The `dirname` is used to specify the folder where persistable variables
+        are going to be saved. If you would like to save variables in separate
+        files, set `filename` None; if you would like to save all variables in a
+        single file, use `filename` to specify the file name.
+
+        Args:
+            mode(Mode): the implementation of Fleet's mode.
+
+        Returns:
+            None
+        """
+
+        assert isinstance(mode, Mode)
+        self.is_initialized = False
+        self.mode = mode
+        self.workers = 0
+        self.servers = 0
+        self.worker_endpoints = []
+        self.server_endpoints = []
+        self.role = Role.WORKER
+        self.current_endpoint = None
+        self.current_id = 0
+        self.optimizer = None
+        self.role_maker_ = None
+
+    def is_first_worker(self):
+        """
+        Check whether the node is the first instance of worker.
+
+        Returns:
+            bool: True if this is the first node of worker,
+                  False if not.
+        """
+        return self.is_worker() and self.current_id == 0
+
+    def worker_id(self):
+        """
+        Get current worker id.
```

Comment:
I think worker_id should be clarified here. In Collective mode, a worker id corresponds a gpu device card. In Parameter Server mode, worker id should be a pod that runs multi-thread training.